File: apps/desktop/lib/core/obsidian/obsidian_writer.dart
import 'dart:io';

import 'package:path/path.dart' as p;

class ObsidianWriteResult {
  const ObsidianWriteResult({required this.markdownPath, required this.fileName});

  final String markdownPath;
  final String fileName;
}

/// 把 Markdown 写入 Obsidian Vault 的笔记目录。纯 dart:io，可跨端复用。
class ObsidianWriter {
  const ObsidianWriter();

  /// 写入 `<vaultPath>/<capturesDir>/<baseName>.md`。
  /// 自动创建目录、清洗非法文件名、重名时追加 (1)/(2)…
  Future<ObsidianWriteResult> writeNote({
    required String vaultPath,
    required String capturesDir,
    required String baseName,
    required String markdown,
  }) async {
    final dirPath = p.normalize(p.join(vaultPath, capturesDir));
    await Directory(dirPath).create(recursive: true);

    final safeBase = sanitizeFileName(baseName);
    var fileName = '$safeBase.md';
    var file = File(p.join(dirPath, fileName));
    var i = 1;
    while (file.existsSync()) {
      fileName = '$safeBase ($i).md';
      file = File(p.join(dirPath, fileName));
      i++;
    }
    await file.writeAsString(markdown, flush: true);
    return ObsidianWriteResult(markdownPath: file.path, fileName: fileName);
  }

  /// 清洗 Windows 非法文件名字符（\ / : * ? " < > | 与控制字符）。
  static String sanitizeFileName(String name) {
    var s = name.replaceAll(RegExp(r'[\\/:*?"<>|\x00-\x1F]'), ' ');
    s = s.replaceAll(RegExp(r'\s+'), ' ').trim();
    s = s.replaceAll(RegExp(r'[ .]+$'), ''); // Windows 不允许结尾的点/空格
    if (s.isEmpty) s = 'capture';
    if (s.length > 80) s = s.substring(0, 80).trim();
    return s;
  }
}

File: apps/macos/Sources/Services/ObsidianWriter.swift
import Foundation

struct ObsidianWriteResult {
    let markdownPath: String
    let fileName: String
}

/// 把 Markdown 写入 Obsidian Vault 的笔记目录。对齐 Windows obsidian_writer.dart。
/// 自动建目录、清洗非法文件名、重名追加 (1)/(2)…
struct ObsidianWriter {

    func writeNote(vaultPath: String, capturesDir: String, baseName: String,
                   markdown: String) throws -> ObsidianWriteResult {
        let dir = URL(fileURLWithPath: vaultPath, isDirectory: true)
            .appendingPathComponent(capturesDir, isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)

        let safe = Self.sanitizeFileName(baseName)
        var fileName = "\(safe).md"
        var fileURL = dir.appendingPathComponent(fileName)
        var i = 1
        while FileManager.default.fileExists(atPath: fileURL.path) {
            fileName = "\(safe) (\(i)).md"
            fileURL = dir.appendingPathComponent(fileName)
            i += 1
        }
        try markdown.write(to: fileURL, atomically: true, encoding: .utf8)
        return ObsidianWriteResult(markdownPath: fileURL.path, fileName: fileName)
    }

    /// 清洗非法文件名字符（\ / : * ? " < > | 与控制字符），跨平台保守处理。
    static func sanitizeFileName(_ name: String) -> String {
        var s = name.replacingOccurrences(
            of: "[\\\\/:*?\"<>|\\x00-\\x1f]", with: " ", options: .regularExpression)
        s = s.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespaces)
        s = s.replacingOccurrences(of: "[ .]+$", with: "", options: .regularExpression)
        if s.isEmpty { s = "capture" }
        if s.count > 80 {
            s = String(s.prefix(80)).trimmingCharacters(in: .whitespaces)
        }
        return s
    }
}
